aws/resource_aws_securityhub_insight.go

```diff
@@ -0,0 +1,1241 @@
+				},
+				"value": {
+					Type:     schema.TypeString,
+					Required: true,
+				},
+			},
+		},
+	}
+}
+
+func numberFilterSchema() *schema.Schema {
+	return &schema.Schema{
+		Type:     schema.TypeSet,
+		Optional: true,
+		MaxItems: 20,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"eq": {
+					Type:         schema.TypeString,
+					Optional:     true,
+					ValidateFunc: validateTypeStringNullableFloat,
+				},
+				"gte": {
+					Type:         schema.TypeString,
+					Optional:     true,
+					ValidateFunc: validateTypeStringNullableFloat,
+				},
+				"lte": {
+					Type:         schema.TypeString,
+					Optional:     true,
+					ValidateFunc: validateTypeStringNullableFloat,
+				},
+			},
+		},
+	}
+}
+
+func stringFilterSchema() *schema.Schema {
+	return &schema.Schema{
+		Type:     schema.TypeSet,
+		Optional: true,
+		MaxItems: 20,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"comparison": {
+					Type:         schema.TypeString,
+					Required:     true,
+					ValidateFunc: validation.StringInSlice(securityhub.StringFilterComparison_Values(), false),
+				},
+				"value": {
+					Type:     schema.TypeString,
+					Required: true,
+				},
+			},
+		},
+	}
+}
+
+func workflowStatusSchema() *schema.Schema {
+	s := stringFilterSchema()
+
+	s.Elem.(*schema.Resource).Schema["value"].ValidateFunc = validation.StringInSlice(securityhub.WorkflowStatus_Values(), false)
+
+	return s
+}
+
+func expandSecurityHubDateFilterDateRange(l []interface{}) *securityhub.DateRange {
+	if len(l) == 0 || l[0] == nil {
+		return nil
+	}
+
+	tfMap, ok := l[0].(map[string]interface{})
+	if !ok {
+		return nil
+	}
+
+	dr := &securityhub.DateRange{}
+
+	if v, ok := tfMap["unit"].(string); ok && v != "" {
+		dr.Unit = aws.String(v)
+	}
+
+	if v, ok := tfMap["value"].(int); ok {
+		dr.Value = aws.Int64(int64(v))
+	}
+
+	return dr
+}
+
+func expandSecurityHubDateFilters(l []interface{}) []*securityhub.DateFilter {
+	if len(l) == 0 || l[0] == nil {
+		return nil
+	}
+
+	var dateFilters []*securityhub.DateFilter
+
+	for _, item := range l {
+		tfMap, ok := item.(map[string]interface{})
+		if !ok {
+			continue
+		}
+
+		df := &securityhub.DateFilter{}
+
+		if v, ok := tfMap["date_range"].([]interface{}); ok && len(v) > 0 && v[0] != nil {
+			df.DateRange = expandSecurityHubDateFilterDateRange(v)
+		}
+
+		if v, ok := tfMap["end"].(string); ok && v != "" {
+			df.End = aws.String(v)
+		}
+
+		if v, ok := tfMap["start"].(string); ok && v != "" {
+			df.Start = aws.String(v)
+		}
+
+		dateFilters = append(dateFilters, df)
+	}
+
+	return dateFilters
+}
+
+func expandSecurityHubSecurityFindingFilters(l []interface{}) *securityhub.AwsSecurityFindingFilters {
+	if len(l) == 0 || l[0] == nil {
+		return nil
+	}
+
+	tfMap, ok := l[0].(map[string]interface{})
+	if !ok {
+		return nil
+	}
+
+	filters := &securityhub.AwsSecurityFindingFilters{}
+
+	if v, ok := tfMap["aws_account_id"].(*schema.Set); ok && v.Len() > 0 {
```

Comment:
Is there any risk with the `tfMap["aws_account_id"].(*schema.Set)` vs. this approach:

```
	if v, ok := tfMap["aws_account_id"]; ok && v.(*schema.Set).Len() > 0 {
 		filters.AwsAccountId = expandSecurityHubStringFilters(v.(*schema.Set).List())
 	}
```

I know for simple types it is okay but not sure if there's anyway `tfMap["aws_account_id"].(*schema.Set)` can error (e.g., nil pointer). I feel more comfortable with this approach only because it seems less likely to panic.

However, if it's equally impossible to panic either way, then your approach is certainly clearer by avoiding the casts in two places. In that case, we should also update the docs: https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/contributing/data-handling-and-conversion.md#root-typeset-of-resource-and-aws-list-of-structure

Reply:
hmm so i _think_ the type casting is ok even if we're trying to unpack as a pointer type
e.g.
https://play.golang.org/p/Y-FE4Nc93og

looks like we're covered https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/contributing/data-handling-and-conversion.md#nested-typeset-of-resource-and-aws-list-of-structure 👍